File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.util.*;
import java.io.*;


public class Main {

    public static final int INF = Integer.MAX_VALUE / 2;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());

        PriorityQueue<Pair> pq = new PriorityQueue<>((o1,o2) -> o1.c - o2.c);
        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            pq.add(new Pair(a, b, c));
        }
        int[] parents = new int[v+1]; // parents[x]: x노드의 부모 노드
        for(int i = 0; i <= v; i++){
            parents[i] = i; // 초기화
        }

        int answer = 0;
        int time = 0;
        while (true){
            if(time == v-1) break;
            Pair cur = pq.poll();
            if(isDiffGroup(cur.a, cur.b, parents)){
                union(cur.a, cur.b, parents);
                answer += cur.c;
                time += 1;
            }
        }
        System.out.println(answer);
    }

    public static int getParent(int v, int[] p){
        if(p[v] == v) return v;

        return p[v] = getParent(p[v], p);
    }

    public static void union(int x, int y, int[] p){
        int p1 = getParent(x, p);
        int p2 = getParent(y, p);
        if(p1 < p2){
            p[p2] = p1;
        }else{
            p[p1] = p2;
        }
    }

    public static boolean isDiffGroup(int x, int y, int[] p){
        int p1 = getParent(x, p);
        int p2 = getParent(y, p);
        return p1 != p2;
    }

    static class Pair{
        int a;
        int b;
        int c;

        public Pair(int a, int b, int c){
            this.a = a;
            this.b = b;
            this.c = c;
        }
    }
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
import heapq

sys.setrecursionlimit(10**9)
input = sys.stdin.readline


class Data:
    def __init__(self, v1, v2, w):
        self.v1 = v1
        self.v2 = v2
        self.w = w

    def __lt__(self, other):
        return self.w < other.w


def get_parent(v, parents):
    if parents[v] == v:
        return v

    parents[v] = get_parent(parents[v], parents)
    return parents[v]


def is_diff_group(v1, v2, parents):
    p1 = get_parent(v1, parents)
    p2 = get_parent(v2, parents)

    return p1 != p2


def union(v1, v2, parents):
    p1 = get_parent(v1, parents)
    p2 = get_parent(v2, parents)
    if p1 < p2:
        parents[p2] = p1
    else:
        parents[p1] = p2


heap = []
v, e = map(int, input().split(' '))

for _ in range(e):
    a, b, c = map(int, input().split(' '))
    heapq.heappush(heap, Data(a, b, c))

parents = [i for i in range(0, v+1)]  # parents[x]: x의 부모 노드. 초기화: 자기 자신 노드

answer = 0
cnt = 0
while cnt != v-1:
    cur = heapq.heappop(heap)

    if is_diff_group(cur.v1, cur.v2, parents):
        union(cur.v1, cur.v2, parents)
        answer += cur.w
        cnt += 1

print(answer)
